# Let's prep our data

### Get the data

- Download to your computer <a href="https://drive.google.com/file/d/1-798fq909hN3x3UxBYUDZrkEaPTliyvJ/view?usp=sharing"> this 256MB file</a> that contains NYPD summons data from 2015-2025s.


In [1]:
## import libraries
import pandas as pd


In [2]:
# import data
nypd_summons = pd.read_csv("summons-15-25.csv")
nypd_summons.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1384334 entries, 0 to 1384333
Data columns (total 18 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   SUMMONS_KEY            1384334 non-null  int64  
 1   SUMMONS_DATE           1384334 non-null  object 
 2   OFFENSE_DESCRIPTION    1384316 non-null  object 
 3   LAW_SECTION_NUMBER     1348831 non-null  object 
 4   LAW_DESCRIPTION        1309406 non-null  object 
 5   SUMMONS_CATEGORY_TYPE  1309321 non-null  object 
 6   AGE_GROUP              1354372 non-null  object 
 7   SEX                    1284382 non-null  object 
 8   RACE                   964769 non-null   object 
 9   JURISDICTION_CODE      1384334 non-null  int64  
 10  BORO                   1372084 non-null  object 
 11  PRECINCT_OF_OCCUR      1384334 non-null  int64  
 12  X_COORDINATE_CD        1383871 non-null  float64
 13  Y_COORDINATE_CD        1383871 non-null  float64
 14  Latitude          

## Doesn't AI make learning coding obsolete?

Yes, it's excellent at data analysis and coding. 

But here are a few problems you might run into:

- How will you investigate AI-related topics?
- Hitting Limits – even with paid accounts.
- Protecting sources.
- Not infallibable.

At Bloomberg, the rule is **you are responsible** for every number, analysis and finding you put into your pieces.

So be prepared to understand and explain each of your decisions, including AI's guidance.

## Traditonal coding v. Vibe Coding


**Traditional coding**
- structured approach to learning syntax and rules first.

**Vibe coding**
- learning by experimenting, observing results, analyzing and iterating – with AI's assistance – rather than memorizing syntax.

- still need to learn coding because you must understand what the code is doing, recognizing mistakes and incorrect interpretations of your prompts.

So you need to retain control of your editorial decisions and process. AI can help write code, but understanding coding helps you trust — or question — the answers.


What are some things - terms, processes etc – that we learned so far?
- Python is good at math
- variables are characters in your data.
- PEMDAS
- f-strings
- built in functions
    - like what? len, type, print, abs, round
- importing and reading data
- understanding dataframes v. series
- getting an overview of your data

# Interviewing Data

If AI can write the code, what questions might you ask based on your domain/beat knowledge?

Let's step back to see how we <a href="https://docs.google.com/presentation/d/1oA50DVvcPT2IWDAHTXogD5lWsWk1TVq_sp5iIS1qQv8/edit?usp=sharing">"interview"</a> data.

## Interviewing Data with AI



In [ ]:
## Get a sample. 
nypd_summons.sample(10)

,SUMMONS_KEY,SUMMONS_DATE,OFFENSE_DESCRIPTION,LAW_SECTION_NUMBER,LAW_DESCRIPTION,SUMMONS_CATEGORY_TYPE,AGE_GROUP,SEX,RACE,JURISDICTION_CODE,BORO,PRECINCT_OF_OCCUR,X_COORDINATE_CD,Y_COORDINATE_CD,Latitude,Longitude,Lon_Lat,YEAR
885252,156752446,2016-09-12,OTHER PENAL LAW,9999,Penal Law,OTHER PENAL LAW,<18,F,WHITE HISPANIC,0,BRONX,43,1017689.0,241175.0,40.828590,-73.879172,POINT (-73.87917218299998 40.82859008200006),2016
100149,278536485,2023-12-05,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,(null),(null),0,BROOKLYN,76,984166.0,188346.0,40.683651,-74.000304,POINT (-74.0003043120176 40.6836512406856),2023
1091887,148709049,2015-12-19,OTHER VTL,9999,VTL,VTL,25-44,M,NaN,0,QUEENS,113,1050551.0,189214.0,40.685785,-73.760941,POINT (-73.76094069799996 40.685785390000035),2015
748082,164183567,2017-04-25,OTHER ADMIN CODE,9999,Administrative Code,OTHER ADMIN CODE,45-64,M,WHITE HISPANIC,0,BRONX,46,1011399.0,250642.0,40.854596,-73.901862,POINT (-73.90186227199997 40.85459579500008),2017
466729,187814429,2018-09-19,CONSUMPTION OF ALCOHOL IN PARK,1-05F,Park Regs,ALCOHOL,25-44,M,WHITE HISPANIC,0,BROOKLYN,72,982967.0,175044.0,40.647140,-74.004625,POINT (-74.00462486 40.6471401),2018
709335,166438629,2017-06-22,"MARIJUANA, POSSESSION OF",221.05,Penal Law,MARIJUANA,18-24,M,BLACK,0,BROOKLYN,77,998567.0,186140.0,40.677585,-73.948385,POINT (-73.94838521299994 40.677584695000064),2017
99131,279332510,2023-12-09,LITTERING,16-118,Administrative Code,LITTERING,25-44,M,BLACK,0,QUEENS,103,1042136.0,193123.0,40.696565,-73.791247,POINT (-73.79124747557337 40.69656543775117),2023
1354373,141153891,2015-02-25,ACCEPT ON HAIL,19-507,Administrative Code,TLC,45-64,M,UNKNOWN,0,QUEENS,105,1056903.0,200871.0,40.717731,-73.737912,POINT (-73.73791173999997 40.71773112700004),2015
1382514,140508991,2015-01-11,NO SECURITY GUARD LICENSE,89-C,General Business Law,BUSINESS,45-64,M,UNKNOWN,0,BROOKLYN,67,997468.0,174911.0,40.646765,-73.952369,POINT (-73.95236938199997 40.64676527900008),2015
1148049,146589507,2015-09-30,VENDOR: UNLICENSED,20-453,Administrative Code,VENDING,45-64,M,NaN,0,BROOKLYN,83,1006188.0,194405.0,40.700255,-73.920883,POINT (-73.92088277299997 40.70025461000005),2015


### What are some questions you could ask this data?

1. Which borough had the most summons?
2. How many times did the top 10 most common offenses occur?
3. What are the total number of summons broken down by year?
4. What were some of the most unusual summons? (how do you define most unusual)
5. What is the count of summons by gender?
6. How many summons were issued by race?
7. Which borough had the fewest offenses per 100K people?

### Upload data to AI - ask it questions.


#### What happened?

What do we do on deadline when the dataset is too big?

Even if AI ingests a large dataset, it responds by analyzing a fraction of the data.

The importance of <a href="https://docs.google.com/presentation/d/1BKn8iUq9wy3KgrcfbfpJN6i3cnB8dNlnbAQlrbUSGJ0/edit?usp=sharing">Reproducible Data Journalism</a>

## And now let's interview data using Pandas



We are soon going to blend using AI and coding, but first it's important to understand some Pandas fundamentals.

#### Finally **Vibe Coding**!

##### Template for Vibe Coding

1. Give AI a sample of your data (like a 100 rows)
2. Ask it for the Pandas code to answer a question, **but always tell it to comment the code**!
3. Leave a comment in Markdown with your prompt
4. Test it out in a code cell.
5. If the code throws an error, copy the entire code and give it AI.
6. Remember to add this updated prompt to your markdown.
7. Test out the the updated code in same cell.
8. Create a markdown cell in which you use f-strings to leave your finding

#### 1. Give AI a sample of your data (like a 100 rows) 

In [8]:
## see sample data
summons_sample = nypd_summons.sample(100)
summons_sample

,SUMMONS_KEY,SUMMONS_DATE,OFFENSE_DESCRIPTION,LAW_SECTION_NUMBER,LAW_DESCRIPTION,SUMMONS_CATEGORY_TYPE,AGE_GROUP,SEX,RACE,JURISDICTION_CODE,BORO,PRECINCT_OF_OCCUR,X_COORDINATE_CD,Y_COORDINATE_CD,Latitude,Longitude,Lon_Lat,YEAR
1332078,141674837,2015-03-25,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,NaN,D,UNKNOWN,0,BRONX,41,1013847.0,237215.0,40.817735,-73.893072,POINT (-73.89307240799997 40.81773472000003),2015
450619,190785884,2018-12-05,DISPLAY OF PEDICAB DRIVER?S IDENTIFICATION,20-258,NaN,NaN,45-64,M,WHITE,0,NEW YORK,18,991326.0,217002.0,40.762302,-73.974458,POINT (-73.97445832 40.76230221),2018
1097002,148500061,2015-12-11,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,25-44,M,NaN,0,QUEENS,110,1024909.0,211402.0,40.746841,-73.853264,POINT (-73.85326351999998 40.746840990000074),2015
257652,239220075,2022-01-16,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,(null),(null),0,BRONX,52,1019342.0,258247.0,40.875435,-73.873106,POINT (-73.873106 40.875435),2022
144058,271636082,2023-07-20,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,(null),(null),0,BRONX,48,1019154.0,243876.0,40.835990,-73.873863,POINT (-73.87386305732169 40.83598959091821),2023
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
143059,271716473,2023-07-22,ALCOHOLIC BEVERAGE IN PUBLIC,10-125(2B),(null),(null),45-64,M,BLACK,2,BROOKLYN,88,991150.0,192509.0,40.695068,-73.975116,POINT (-73.975116 40.695068),2023
1202507,145188405,2015-08-07,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,25-44,M,NaN,0,BROOKLYN,79,1000754.0,192580.0,40.695257,-73.940485,POINT (-73.94048477899997 40.69525721300005),2015
340722,211422882,2020-03-24,FEDERAL MOTOR VEH. SAFETY REG,NaN,NYS Transportation,NYS TRANS,UNKNOWN,NaN,NaN,0,BROOKLYN,84,989876.0,189633.0,40.687182,-73.979715,POINT (-73.97971544 40.68718198),2020
738757,164552081,2017-05-07,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,25-44,M,WHITE HISPANIC,0,BRONX,46,1009701.0,249613.0,40.851777,-73.908004,POINT (-73.90800416599996 40.85177655500007),2017


In [ ]:
# Summon the Head 
nypd_summons.head()

,SUMMONS_KEY,SUMMONS_DATE,OFFENSE_DESCRIPTION,LAW_SECTION_NUMBER,LAW_DESCRIPTION,SUMMONS_CATEGORY_TYPE,AGE_GROUP,SEX,RACE,JURISDICTION_CODE,BORO,PRECINCT_OF_OCCUR,X_COORDINATE_CD,Y_COORDINATE_CD,Latitude,Longitude,Lon_Lat,YEAR
0,298767983,2024-12-31,DISORDERLY CONDUCT,240.2,Penal Law,DISORDERLY CONDUCT,45-64,M,WHITE HISPANIC,0,MANHATTAN,14,985927.0,213077.0,40.751523,-73.993947,POINT (-73.99394730218468 40.75152348693427),2024
1,298704187,2024-12-31,IMPROPER PLATES,402,VTL,VTL,18-24,M,ASIAN / PACIFIC ISLANDER,0,BRONX,49,1020823.0,249562.0,40.851589,-73.867801,POINT (-73.86780058180116 40.85158922344624),2024
2,298787872,2024-12-31,TRESPASS,140.05,Penal Law,TRESPASS,25-44,F,BLACK HISPANIC,2,MANHATTAN,32,998922.0,235604.0,40.813352,-73.946996,POINT (-73.946996 40.813352),2024
3,298724988,2024-12-31,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,(null),(null),0,BROOKLYN,94,1001547.0,203048.0,40.723980,-73.937597,POINT (-73.93759674847603 40.72397961334044),2024
4,298704460,2024-12-31,POSSESSION OF ALCOHOL BY A MINOR,65-C,(null),(null),18-24,M,WHITE,0,MANHATTAN,6,984439.0,204595.0,40.728254,-73.999318,POINT (-73.999318 40.728254),2024


In [10]:
## export
summons_sample.to_csv("sample_summons_100.csv", index = False, encoding = "UTF8")

### Explanation of the export code

`summons_sample` is the dataframe we want to export

`.to_csv()` saves our dataframe to a .csv file. You are baseically exporting your python data out into a format that Excel, Google Sheets can read.

`index=False`
By default, on export pandas includes the DataFrame's index – the row numbers (0, 1, 2, 3...). We NEVER
need it 99.99 percent of the time. `index=False` simply tells pandas to leave that column out on export.

`encoding="utf-8"`
Tells Python which character encoding to use when writing the file. `UTF-8` is the most universal standard — it handles English plus accented characters, symbols, non-Latin scripts, etc. without garbling them. If your data has names like "García" or "Müller" and you skip this, you might get weird characters or errors depending on your system.

Recall some questions we want to ask:

1. Which borough had the most summons?
2. How many times did the top 10 most common offenses occur?
3. What are the total number of summons broken down by year?
4. What were some of the most unusual summons? (how do you define most unusual)
5. What is the count of summons by gender?
6. How many summons were issued by race?
7. Which borough had the fewest offenses per 100K people?

## Who wants to share their answers?

# Prompt for all questions #
For all the following questions, use Pandas 2.5 or newer.
Always provide comments for each line of code. 




### Prompt: 

Provide the pandas code to calculate which boroughs had the most summons (in descending order).


In [12]:
# 1. Which borough had the most summons? 

# Replace string '(null)' placeholders with real NaN so they're excluded cleanly
nypd_summons['BORO'] = nypd_summons['BORO'].replace('(null)', pd.NA)

# Count summons per borough, sort descending, and drop any missing borough values
borough_counts = (
    nypd_summons['BORO']
    .value_counts()              # Count occurrences of each borough; sorts descending by default
    .reset_index()               # Convert the Series into a two-column DataFrame
    .rename(columns={            # Rename columns from pandas defaults to readable names
        'BORO': 'borough',
        'count': 'summons_count'
    })
)

borough_counts


,borough,summons_count
0,BROOKLYN,427679
1,BRONX,312327
2,MANHATTAN,275884
3,QUEENS,246314
4,NEW YORK,63056
5,STATEN ISLAND,46824


## I'm using Claude to answer the following questions. 

These are the basic instructions I gave: 

_Hi this is the name of my actual data : nypd_summons = pd.read_csv("summons-15-25.csv")
but the file is to big, so I'll upload just a sample here and will ask you to help me write code for data analysis. For all the following questions, use Pandas 2.5 or newer. Always provide comments for each line of code._

### Prompt #2: 

**Provide the pandas code to calculate what the top 10 most common offenses were (in descending order). Ask me clarifying questions until you are 95% sure you can provide a perfect result.**

Two quick questions before I write the code:

1. **Which column should define "offense"?**
   - `OFFENSE_DESCRIPTION` — granular (40+ unique values in the sample)
   - `SUMMONS_CATEGORY_TYPE` — broader groupings (~20 categories)
   - Both — show results for each

2. **How should the results be displayed?**
   - Raw count only
   - Raw count + percentage of total
   - Raw count + percentage + cumulative percentage

Q: Which column should be used to define 'offense'?
A: OFFENSE_DESCRIPTION (granular, 40+ unique values)

Q: How should the count be displayed? (Select all that apply)
A: Raw count

In [12]:
# 2. What were the top 10 most common offenses?

# Load the dataset
nypd_summons = pd.read_csv("summons-15-25.csv")

# Count occurrences of each unique offense description,
# sort in descending order, and keep only the top 10
top_10_offenses = (
    nypd_summons["OFFENSE_DESCRIPTION"]  # Select the offense description column
    .value_counts()                       # Count each unique value, sorted descending by default
    .head(10)                             # Keep only the top 10
    .rename_axis("OFFENSE_DESCRIPTION")  # Name the index column explicitly
    .reset_index(name="COUNT")            # Convert index to column, name the count column "COUNT"
)

# Display the result
top_10_offenses

,OFFENSE_DESCRIPTION,COUNT
0,CONSUMPTION OF ALCOHOL,248650
1,FEDERAL MOTOR VEH. SAFETY REG,158420
2,"MARIJUANA, POSSESSION OF",84910
3,URINATING IN PUBLIC,50340
4,RECKLESS DRIVING,44691
5,ALCOHOLIC BEVERAGE IN PUBLIC,43789
6,FAILURE TO COMPLY WITH SIGN,42810
7,TRESPASS,38595
8,OTHER ADMIN CODE,31851
9,DISORDERLY CONDUCT,20743


### Findings #2: 

Alcohol consumption, not following Federal Motor Vehicle Safety Regulations, and marijuana possession were the most common offenses between 2015 and 2025, followed by urinating in public, reckless driving, drinking alcohol in public, not following a street sign, and trespassing. It is unclear what "Other Admin Code" implies. Disorderly conduct is ranked the tenth most common offense, with about 20,700 summons between 2015 and 2025.

### Prompt #3: 

**Provide the pandas code to calculate the total number of summons broken down by year. Please ask me clarifying questions until you are sure you can provide a perfect result.**

Two quick questions:

1. **How should the years be sorted?**
   - Ascending (2015 → 2025)
   - Descending (2025 → 2015)

2. **How should the results be displayed?**
   - Raw count only
   - Raw count + percentage of total
   - Raw count + year-over-year change
   - All three

Q: How should the years be sorted?
A: Ascending (2015 → 2025)

Q: How should the results be displayed? (Select all that apply)
A: Raw count


In [14]:
# 3. What are the total number of summons broken down by year?

# Group the data by the YEAR column and count the number of summons per year,
# then reset the index to convert the result into a clean DataFrame
summons_by_year = (
    nypd_summons.groupby("YEAR")          # Group rows by year
    .size()                                # Count the number of rows per group
    .reset_index(name="COUNT")             # Convert index to column, name the count column "COUNT"
    .sort_values("YEAR", ascending=True)   # Sort by year in ascending order (2015 → 2025)
    .reset_index(drop=True)                # Reset index after sorting for clean output
)

# Display the result
summons_by_year


,YEAR,COUNT
0,2015,297748
1,2016,267950
2,2017,165093
3,2018,89895
4,2019,85899
5,2020,55589
6,2021,45717
7,2022,58059
8,2023,106529
9,2024,94542


### Findings #3:

In 2015, almost 300,000 summons were issued, whereas in 2025 only about 117,000 were issued. There was a notable dip in 2021, during the pandemic, with only 45,700 summons. While numbers were declining before the pandemic and rising again after 2021, summons levels never reached pre-pandemic numbers again.

### Prompt #4: 

**Give me the pandas code to calculate the most unusual summons. Please ask me clarifying questions until you can provide a perfect result.**


"Most unusual" can mean a few different things with this data — let me narrow it down!

What does "most unusual" mean to you?

Rarest offenses — lowest total count overall
Offenses that spiked or dropped sharply in a specific year
Offenses unique to a single borough
Offenses issued to unusual demographic combinations (age, sex, race)
Or did you have something else in mind entirely?


**Let's define "most unusual" as offenses issued to unusual demographic combinations (age, sex, race)for now.** 


Great direction! A few more questions to nail this down:

Three quick questions:

Which demographic columns should be included?

AGE_GROUP, SEX, RACE (or any combination)

How should "unusual" be defined?

Rarest combinations (lowest count)
Most overrepresented vs. overall population share
Combinations that appear only once

How should missing values be handled? (RACE has ~36% nulls in the sample)

Exclude rows with nulls
Keep them and label as "UNKNOWN"


**Let's include age, sex and race and see what the rarest combinations are. Keep the rows with nulls and label them as unknown.**

In [16]:
# 4. What were some of the most unusual summons?

# Fill missing values in demographic columns with 'UNKNOWN'
nypd_summons[["SEX", "AGE_GROUP", "RACE"]] = nypd_summons[
    ["SEX", "AGE_GROUP", "RACE"]
].fillna("UNKNOWN")                                        # Replace NaN with 'UNKNOWN' in all three columns

# Group by all three demographic columns and count occurrences,
# then reset the index to convert the result into a clean DataFrame
unusual_demographics = (
    nypd_summons.groupby(["SEX", "AGE_GROUP", "RACE"])    # Group by the three demographic columns
    .size()                                                # Count the number of rows per group
    .reset_index(name="COUNT")                             # Convert index to column, name it "COUNT"
    .sort_values("COUNT", ascending=True)                  # Sort ascending — rarest combinations first
    .reset_index(drop=True)                                # Reset index after sorting for clean output
)

# Display the result
unusual_demographics




,SEX,AGE_GROUP,RACE,COUNT
0,M,418,UNKNOWN,1
1,M,243,UNKNOWN,1
2,M,240,UNKNOWN,1
3,M,237,BLACK,1
4,M,229,UNKNOWN,1
...,...,...,...,...
283,M,18-24,UNKNOWN,91621
284,UNKNOWN,UNKNOWN,UNKNOWN,99952
285,M,45-64,UNKNOWN,103683
286,M,25-44,BLACK,132615


### Prompt #4A:

**I wasn't clear enough, I of course also want to know what the offense was, not only how rare a summon occured. Give me the pandas code to calculate the Top 20 most unusual summons and offenses. Ask me clarifying questions if needed.**

Q: Which offense column should be included?
A: OFFENSE_DESCRIPTION (granular)

In [20]:
# 4. What were the Top20 most unusual offenses by Race, Gender and Age? 

# Fill missing values in demographic columns with 'UNKNOWN'
nypd_summons[["SEX", "AGE_GROUP", "RACE"]] = nypd_summons[
    ["SEX", "AGE_GROUP", "RACE"]
].fillna("UNKNOWN")                                                      # Replace NaN with 'UNKNOWN' in all three columns

# Group by offense and all three demographic columns and count occurrences,
# then reset the index to convert the result into a clean DataFrame
top_20_unusual = (
    nypd_summons.groupby(                                                # Group by offense + demographics
        ["OFFENSE_DESCRIPTION", "SEX", "AGE_GROUP", "RACE"]
    )
    .size()                                                              # Count the number of rows per group
    .reset_index(name="COUNT")                                           # Convert index to column, name it "COUNT"
    .sort_values("COUNT", ascending=True)                                # Sort ascending — rarest combinations first
    .head(20)                                                            # Keep only the top 20 rarest combinations
    .reset_index(drop=True)                                              # Reset index after sorting for clean output
)

# Display the result
top_20_unusual

,OFFENSE_DESCRIPTION,SEX,AGE_GROUP,RACE,COUNT
0,20-217(I),F,18-24,BLACK,1
1,NO TOW LICENSE,(null),(null),BLACK,1
2,NO TOW LICENSE,F,25-44,WHITE HISPANIC,1
3,NO TOW LICENSE,M,18-24,ASIAN / PACIFIC ISLANDER,1
4,NO TOW LICENSE,M,25-44,AMERICAN INDIAN/ALASKAN NATIVE,1
5,NO TOW LICENSE,M,25-44,ASIAN / PACIFIC ISLANDER,1
6,NO TOW LICENSE,M,25-44,UNKNOWN,1
7,NO TOW LICENSE,(null),(null),ASIAN / PACIFIC ISLANDER,1
8,NO TOW LICENSE,M,45-64,BLACK HISPANIC,1
9,NO TOW LICENSE,M,<18,WHITE HISPANIC,1


### Prompt 4B: 

okay could you also give me the code to calculate the most unusual summons in terms of top10 offenses that occurred the least?

In [21]:
# 4. What were the Top10 offenses that occured the least? 

# Count the total number of summons per offense description,
# sort ascending to get the rarest offenses first, and keep only the top 10
top_10_rarest_offenses = (
    nypd_summons["OFFENSE_DESCRIPTION"]                  # Select the offense description column
    .value_counts()                                       # Count each unique value, sorted descending by default
    .tail(10)                                             # Keep the 10 lowest counts (rarest offenses)
    .rename_axis("OFFENSE_DESCRIPTION")                  # Name the index column explicitly
    .reset_index(name="COUNT")                            # Convert index to column, name it "COUNT"
    .sort_values("COUNT", ascending=True)                 # Sort ascending — rarest offenses first
    .reset_index(drop=True)                               # Reset index after sorting for clean output
)

# Display the result
top_10_rarest_offenses



,OFFENSE_DESCRIPTION,COUNT
0,ALLOW CONSUMPTION BETWEEN 4:00 AM - 12 NOON (S...,1
1,TRANSFER WITHOUT PERMISSION,1
2,RECEPTACLES; REMOVING FROM STREET AFTER GARBAG...,1
3,ANIMAL/DOG IN PARK BUILDING/BATH FACILITY,1
4,MORE THAN ONE; APPLY FOR/ACCEPT WITHOUT WRITTE...,1
5,ACCIDENT WITH INJURY/DAMAGE; FAIL TO STOP AND ...,1
6,DRIVE THRU BARRIER,1
7,NO PARKING (STREET CLEANING),1
8,METER (BROKEN),1
9,BICYCLE NO LIGHTS,1


### Prompt 4C:

ok don't make it top 10 but any offense that occured only once

In [24]:
# 4. Calculate all offenses that occured only once and therefore are unusual

# Count the total number of summons per offense description
offense_counts = (
    nypd_summons["OFFENSE_DESCRIPTION"]                  # Select the offense description column
    .value_counts()                                       # Count each unique value, sorted descending by default
    .rename_axis("OFFENSE_DESCRIPTION")                  # Name the index column explicitly
    .reset_index(name="COUNT")                            # Convert index to column, name it "COUNT"
)

# Filter to keep only offenses that occurred exactly once
offenses_once = (
    offense_counts[offense_counts["COUNT"] == 1]         # Keep only rows where count equals 1
    .sort_values("OFFENSE_DESCRIPTION", ascending=True)  # Sort alphabetically for easy reading
    .reset_index(drop=True)                               # Reset index after filtering for clean output
)

# Display the result
offenses_once

,OFFENSE_DESCRIPTION,COUNT
0,ACCIDENT WITH INJURY/DAMAGE; FAIL TO STOP AND ...,1
1,ALLOW CONSUMPTION BETWEEN 4:00 AM - 12 NOON (S...,1
2,ANIMAL/DOG IN PARK BUILDING/BATH FACILITY,1
3,AUDIO EQUIPMENT; FAIL TO OBLIGE PASSENGER REGA...,1
4,BICYCLE NO BELL,1
5,BICYCLE NO LIGHTS,1
6,BICYCLE ON GRASS/BRIDLE PATH; RIDE ON,1
7,"CART MAN, UNLICENSED",1
8,"COMMUNICATION SYSTEM, BASE; UNLICENSED",1
9,DANGEROUS CONDITION; CREATE IN PARK/ROAD,1


### Findings 4: 

Some of the most unusual summons that stood out to me were "EMPLOYER ALLOW FEMALE EMPLOYEE TOPLESS", "NO SANITARY SOAP DISPENSER", "TOPLESS FEMALE EMPLOYEE NOT ON STAGE" and "UNREASONABLE NOISE (FROM A MUSICAL INSTRUMENT)".

### Prompt 5:

**Provide the pandas code to calculate the summons by gender. Please ask me clarifying questions until you are sure you can provide a perfect result.**

Q: How should the results be displayed? (Select all that apply)
A: Raw count

Q: How should missing SEX values be handled?
A: No — keep them and label as 'UNKNOWN'

In [26]:
# 5. What is the count of summons by gender?

# Fill missing values in the SEX column with 'UNKNOWN'
nypd_summons["SEX"] = nypd_summons["SEX"].fillna("UNKNOWN")  # Replace NaN with 'UNKNOWN'

# Count the number of summons per gender,
# then reset the index to convert the result into a clean DataFrame
summons_by_gender = (
    nypd_summons.groupby("SEX")                          # Group rows by gender
    .size()                                               # Count the number of rows per group
    .reset_index(name="COUNT")                            # Convert index to column, name it "COUNT"
    .sort_values("COUNT", ascending=False)                # Sort descending — most common gender first
    .reset_index(drop=True)                               # Reset index after sorting for clean output
)

# Display the result
summons_by_gender


,SEX,COUNT
0,M,1013591
1,F,130370
2,(null),105592
3,UNKNOWN,99952
4,D,29961
5,U,4868


### Findings 5: 
Men receive **by far** more summons than women. There is a significant data gap in the number of summons issued to people whose gender is unknown.

### Prompt 6: 
**Provide the pandas code to calculate the summons by race. Please ask me clarifying questions until you are sure you can provide a perfect result.**

Q: How should the results be displayed? (Select all that apply)
A: Raw count

Q: How should missing RACE values be handled?
A: No — keep them and label as 'UNKNOWN'

In [27]:
# 6. How many summons were issued by race?

# Fill missing values in the RACE column with 'UNKNOWN'
nypd_summons["RACE"] = nypd_summons["RACE"].fillna("UNKNOWN")  # Replace NaN with 'UNKNOWN'

# Count the number of summons per race,
# then reset the index to convert the result into a clean DataFrame
summons_by_race = (
    nypd_summons.groupby("RACE")                         # Group rows by race
    .size()                                               # Count the number of rows per group
    .reset_index(name="COUNT")                            # Convert index to column, name it "COUNT"
    .sort_values("COUNT", ascending=False)                # Sort descending — most common race first
    .reset_index(drop=True)                               # Reset index after sorting for clean output
)

# Display the result
summons_by_race


,RACE,COUNT
0,UNKNOWN,605434
1,BLACK,316676
2,WHITE HISPANIC,192409
3,WHITE,87448
4,BLACK HISPANIC,70375
5,(null),54376
6,ASIAN / PACIFIC ISLANDER,49862
7,AMERICAN INDIAN/ALASKAN NATIVE,5735
8,OTHER,2019


### Findings 6: 

For the largest number of summons, race is not documented, which is a data gap that could conceal racial discrimination. Black people and white Hispanics receive by far more summons than white people, who are followed by Black Hispanics, Asians and Pacific Islanders. American Indians and Alaskan Natives account for only about 5,735 summons compared to 316,676 issued to Black people.

# 2020 Census population
```python
populations = {
    "BROOKLYN": 2736074,
    "QUEENS": 2405464,
    "MANHATTAN": 1694251,
    "BRONX": 1472654,
    "STATEN ISLAND": 495747
}
```

### Prompt 7

Provide the pandas code to calculate which borough hat the fewest offenses per 100,000 people. These are the most recent population numbers from the 2020 census. populations = {
    "BROOKLYN": 2736074,
    "QUEENS": 2405464,
    "MANHATTAN": 1694251,
    "BRONX": 1472654,
    "STATEN ISLAND": 495747
}
Please ask me clarifiying questions until you are sure to be able to provide a correct code.

In [28]:
#7. Which borough had the fewest offenses per 100K people?

# Define the 2020 census population for each of the 5 boroughs
populations = {
    "BROOKLYN": 2736074,
    "QUEENS": 2405464,
    "MANHATTAN": 1694251,
    "BRONX": 1472654,
    "STATEN ISLAND": 495747
}

# Filter the dataset to only include rows from the 5 boroughs,
# excluding nulls and any entries not in the populations dictionary
nypd_summons_boroughs = nypd_summons[                        # Filter the original DataFrame
    nypd_summons["BORO"].isin(populations.keys())            # Keep only rows matching the 5 borough names
]

# Count the number of summons per borough,
# then reset the index to convert the result into a clean DataFrame
summons_by_borough = (
    nypd_summons_boroughs.groupby("BORO")                    # Group rows by borough
    .size()                                                   # Count the number of rows per group
    .reset_index(name="COUNT")                                # Convert index to column, name it "COUNT"
)

# Map the population for each borough from the populations dictionary
summons_by_borough["POPULATION"] = summons_by_borough["BORO"].map(populations)  # Add population column

# Calculate the number of summons per 100,000 people for each borough
summons_by_borough["SUMMONS_PER_100K"] = (
    summons_by_borough["COUNT"] /                            # Divide summons count by population
    summons_by_borough["POPULATION"] * 100000                # Multiply by 100,000 to normalize
).round(2)                                                   # Round to 2 decimal places for readability

# Sort by summons per 100,000 ascending — fewest offenses per capita first
summons_by_borough = (
    summons_by_borough
    .sort_values("SUMMONS_PER_100K", ascending=True)         # Sort ascending — fewest per capita first
    .reset_index(drop=True)                                   # Reset index after sorting for clean output
)

# Display the result
summons_by_borough

,BORO,COUNT,POPULATION,SUMMONS_PER_100K
0,STATEN ISLAND,46824,495747,9445.14
1,QUEENS,246314,2405464,10239.77
2,BROOKLYN,427679,2736074,15631.12
3,MANHATTAN,275884,1694251,16283.54
4,BRONX,312327,1472654,21208.44


### Finding 7:

Staten Island had the fewest summons issued per 100,000 people between 2015 and 2025, with about 9,400 summons per 100,000 residents.